sdk/samples/iot/sample_pnp.c

```diff
@@ -0,0 +1,428 @@
+          component_name,
+          az_span_init((uint8_t*)scratch_buf, len),
+          &(json_reader->token),
+          version,
+          context_ptr);
+    }
+
+    if (json_reader->token.kind == AZ_JSON_TOKEN_BEGIN_OBJECT)
+    {
+      if (az_failed(az_json_reader_skip_children(json_reader)))
+      {
+        printf("Failed to skip children of object\r\n");
+        return AZ_ERROR_UNEXPECTED_CHAR;
+      }
+    }
+    else if (json_reader->token.kind == AZ_JSON_TOKEN_END_OBJECT)
+    {
+      break;
+    }
+  }
+
+  return AZ_OK;
+}
+
+// Move reader to the value of property name
+static az_result sample_json_child_token_move(az_json_reader* json_reader, az_span property_name)
+{
+  while (az_succeeded(az_json_reader_next_token(json_reader)))
+  {
+    if ((json_reader->token.kind == AZ_JSON_TOKEN_PROPERTY_NAME)
+        && az_json_token_is_text_equal(&(json_reader->token), property_name))
+    {
+      if (az_failed(az_json_reader_next_token(json_reader)))
+      {
+        printf("Failed to read next token\r\n");
+        return AZ_ERROR_UNEXPECTED_CHAR;
+      }
+
+      return AZ_OK;
+    }
+    else if (json_reader->token.kind == AZ_JSON_TOKEN_BEGIN_OBJECT)
+    {
+      if (az_failed(az_json_reader_skip_children(json_reader)))
+      {
+        printf("Failed to skip child of complex object\r\n");
+        return AZ_ERROR_UNEXPECTED_CHAR;
+      }
+    }
+    else if (json_reader->token.kind == AZ_JSON_TOKEN_END_OBJECT)
+    {
+      return AZ_ERROR_ITEM_NOT_FOUND;
+    }
+  }
+
+  return AZ_ERROR_ITEM_NOT_FOUND;
+}
+
+// Check if the component name is in the model
+static az_result is_component_in_model(
+    az_span component_name,
+    const az_span** sample_components_ptr,
+    int32_t sample_components_num,
+    int32_t* out_index)
+{
+  int32_t index = 0;
+
+  if (az_span_ptr(component_name) == NULL || az_span_size(component_name) == 0)
+  {
+    return AZ_ERROR_UNEXPECTED_CHAR;
+  }
+
+  while (index < sample_components_num)
+  {
+    if ((az_span_size(component_name) == az_span_size(*sample_components_ptr[index]))
+        && (memcmp(
+                (void*)az_span_ptr(component_name),
+                (void*)az_span_ptr(*sample_components_ptr[index]),
+                (size_t)az_span_size(component_name))
+            == 0))
+    {
+      *out_index = index;
+      return AZ_OK;
+    }
+
+    index++;
+  }
+
+  return AZ_ERROR_UNEXPECTED_CHAR;
+}
+
+// Get the telemetry topic for PnP
+az_result pnp_get_telemetry_topic(
+    az_iot_hub_client const* client,
+    az_iot_hub_client_properties* properties,
+    az_span component_name,
+    char* mqtt_topic,
+    size_t mqtt_topic_size,
+    size_t* out_mqtt_topic_length)
+{
+  az_result result;
+
+  az_iot_hub_client_properties pnp_properties;
+
+  if (properties == NULL)
+  {
+    AZ_RETURN_IF_FAILED(az_iot_hub_client_properties_init(
+        &pnp_properties, AZ_SPAN_FROM_BUFFER(pnp_properties_buffer), 0));
+  }
+
+  if ((result = az_iot_hub_client_properties_append(
```

Comment:
Use `az_failed` and `az_succeeded` methods rather than direct comparison to AZ_OK. We may add more success conditions in the future (in fact, we currently have one that is unused, AZ_CONTINUE).

This applies across the board and is the documented recommendation.
https://github.com/Azure/azure-sdk-for-c/blob/df4b7e032abc8adda8e1c27026d2800fbd76c178/sdk/docs/core/README.md#function-results

Reply:
I think I had a discussion with Cris about this before. Whether we do a direct comparison or not, were an additional success case to be added, the resulting behavior might not be as the user expects. Lets assume this flow expects an `AZ_OK` and were `AZ_CONTINUE` to be added, it would want to take a different action: `az_succeeded` would cause unintended behavior. I'm not necessarily disagreeing about changing it to `az_succeeded`, but as long as we recognize that this doesn't necessarily give "more correct" behavior.

Reply:
> but as long as we recognize that this doesn't necessarily give "more correct" behavior.

Fair. Do you agree that using `az_failed` make sense over comparing to a specific failure condition if the only intention is to check whether something failed or succeeded?

In theory, `az_succeeded` is just `!az_failed` and vice versa.

Also, for the sake of argument, let's say we add AZ_SUCCESS `az_result` tomorrow and change all APIs to return that to mean success (not that we would), now the samples that relied on these hard coded comparisons would break.

Reply:
Oh I patched them up. I agree it's cleaner and probably the better way to go. I just wanted to make sure we are aware of the side effects going one way or another.